plugins/volumeFilter_test.go

```diff
@@ -145,3 +146,276 @@ func TestMakeFilterVolume(t *testing.T) {
+			otbBaseCap:         pointy.Float64(0.0),
+			otbQuoteCap:        pointy.Float64(0.0),
+			tbbBaseCap:         pointy.Float64(0.0),
+			tbbQuoteCap:        pointy.Float64(0.0),
+			price:              "2.0",
+			inputAmount:        "100.0",
+			wantAmount:         "",
+			wantTbbBaseCap:     pointy.Float64(0.0),
+			wantTbbQuoteCap:    pointy.Float64(0.0),
+		},
+		{
+			name:               "selling, quote units sell cap, don't keep selling quote, ignore mode",
+			mode:               volumeFilterModeIgnore,
+			sellBaseCapInBase:  nil,
+			sellBaseCapInQuote: pointy.Float64(0),
+			otbBaseCap:         pointy.Float64(0.0),
+			otbQuoteCap:        pointy.Float64(0.0),
+			tbbBaseCap:         pointy.Float64(0.0),
+			tbbQuoteCap:        pointy.Float64(0.0),
+			price:              "2.0",
+			inputAmount:        "100.0",
+			wantAmount:         "",
+			wantTbbBaseCap:     pointy.Float64(0.0),
+			wantTbbQuoteCap:    pointy.Float64(0.0),
+		},
+		{
+			name:               "selling, quote units sell cap, keep selling quote, new amount, exact mode",
+			mode:               volumeFilterModeExact,
+			sellBaseCapInBase:  nil,
+			sellBaseCapInQuote: pointy.Float64(1.0),
+			otbBaseCap:         pointy.Float64(0.0),
+			otbQuoteCap:        pointy.Float64(0.0),
+			tbbBaseCap:         pointy.Float64(0.0),
+			tbbQuoteCap:        pointy.Float64(0.0),
+			price:              "2.0",
+			inputAmount:        "100.0",
+			wantAmount:         "0.5000000",
+			wantTbbBaseCap:     pointy.Float64(0.5),
+			wantTbbQuoteCap:    pointy.Float64(1.0),
+		},
+		{
+			name:               "selling, quote units sell cap, keep selling quote, new amount, ignore mode",
+			mode:               volumeFilterModeIgnore,
+			sellBaseCapInBase:  nil,
+			sellBaseCapInQuote: pointy.Float64(1.0),
+			otbBaseCap:         pointy.Float64(0.0),
+			otbQuoteCap:        pointy.Float64(0.0),
+			tbbBaseCap:         pointy.Float64(0.0),
+			tbbQuoteCap:        pointy.Float64(0.0),
+			price:              "2.0",
+			inputAmount:        "100.0",
+			wantAmount:         "",
+			wantTbbBaseCap:     pointy.Float64(0.0),
+			wantTbbQuoteCap:    pointy.Float64(0.0),
+		},
+		{
+			name:               "selling, quote units sell cap, keep selling quote, exact mode",
+			mode:               volumeFilterModeExact,
+			sellBaseCapInBase:  nil,
+			sellBaseCapInQuote: pointy.Float64(1.0),
+			otbBaseCap:         pointy.Float64(0.0),
+			otbQuoteCap:        pointy.Float64(0.0),
+			tbbBaseCap:         pointy.Float64(0.0),
+			tbbQuoteCap:        pointy.Float64(0.0),
+			price:              "2.0",
+			inputAmount:        "100.0",
+			wantAmount:         "0.5000000",
+			wantTbbBaseCap:     pointy.Float64(0.5),
+			wantTbbQuoteCap:    pointy.Float64(1.0),
+		},
+		{
+			name:               "selling, quote units sell cap, keep selling quote, ignore mode",
+			mode:               volumeFilterModeIgnore,
+			sellBaseCapInBase:  nil,
+			sellBaseCapInQuote: pointy.Float64(1.0),
```

Comment:
I don't fully understand this comment. I get the general importance of precision, but not the rest - I've outlined specific questions below.

I think 1.0 and 0.0 are useful because there's a good amount of math in the function. For example, we add three different fields to calculate `projectedSoldInBaseUnits`. Having simple values makes it easy to debug failures and recalculate possible changes. Having arbitrary values will make that harder.

1. In the context of this function, why would some value be better for testing than another? Ones and zeros might be trivial in general, but I don't understand that in this context. Changing them to `2.34` or `5.78` doesn't test anything new. The only thing I can think about is boundary conditions, but that's separate from the triviality of values.
2. "We actively modify the numbers of offers" - where is this happening? I might be missing something here entirely, but this seems outside the scope of the function we are testing.
3. How would this function affect the precision of anything? Precision could matter where `opToReturn.Amount` is defined. But there, the `fmt.Sprintf` guarantees that `newAmountBeingSold` is converted to a string with 7 decimal places of precision. And the rest of the operations are just addition.

I'm definitely not opposed to changing these values, but I do want to choose meaningful values that increase test coverage, rather than making test failures harder to debug.

Reply:
1. the reason I suggest not using 0.0 and 1.0 is because 0.0 is a nop for addition and 1.0 is a nop for multiplication. so in general those numbers should be avoided when testing math based functions. instead we can use values such as 0.5 and 1.5 for example. 

as an example, if our math in the function does the equivalent of this:
```
add(a, b) {
    return a + b + b
}
```
then `assertEqual(1.0, add(1.0, 0.0))` will be `true` even though the function is broken.
so it's ok to have this in conjunction with another test case such as `assertEqual(1.5, add(1.0, 0.5))`, which will correctly fail. Similarly for multiply when using the value `1.0`

2. sorry, I meant "numbers in the offers".

3. since the final value used in manage offer is a string representation and we are testing the op, then we will automatically test the precision as well. what i meant was that if we test something like `1.001 * 0.12105`, what is the result from rounding, `0.1211710` (round down) or `0.1211711` (round up)?
the test should capture it whatever it is. Then we need to think about what do we want it to be, round up or round down? then we modify the code accordingly and update the test. we can have just 1-2 special test cases for this calling them the "rounding" checks so the remaining tests use simpler numbers.

Reply:
Awesome, thanks for the super thorough reply! This is a really helpful framework for thinking about this. Agreed that we can have a couple tests for the rounding checks.